src/whitenoise/database/accounts_groups.rs

```diff
@@ -0,0 +1,423 @@
+    /// - Attempts to insert first
+    /// - On unique constraint violation, fetches the existing record
+    pub(crate) async fn find_or_create(
+        account_pubkey: &PublicKey,
+        mls_group_id: &GroupId,
+        database: &Database,
+    ) -> Result<(Self, bool), sqlx::Error> {
+        // Try to create first - this handles the race condition properly
+        match Self::create(account_pubkey, mls_group_id, database).await {
+            Ok(created) => Ok((created, true)),
+            Err(sqlx::Error::Database(db_err)) if db_err.is_unique_violation() => {
+                // Another task inserted the record between our check and insert
+                // Fetch the existing record
+                let existing =
+                    Self::find_by_account_and_group(account_pubkey, mls_group_id, database)
+                        .await?
+                        .ok_or(sqlx::Error::RowNotFound)?;
+                Ok((existing, false))
+            }
+            Err(e) => Err(e),
+        }
+    }
+
+    /// Finds all visible AccountGroups for a given account.
+    /// Visible means: user_confirmation is NULL (pending) or true (accepted).
+    /// Declined groups (user_confirmation = false) are hidden.
+    pub(crate) async fn find_visible_for_account(
+        account_pubkey: &PublicKey,
+        database: &Database,
+    ) -> Result<Vec<Self>, sqlx::Error> {
+        let rows = sqlx::query_as::<_, AccountGroupRow>(
+            "SELECT id, account_pubkey, mls_group_id, user_confirmation, created_at, updated_at
+             FROM accounts_groups
+             WHERE account_pubkey = ? AND (user_confirmation IS NULL OR user_confirmation = 1)",
+        )
+        .bind(account_pubkey.to_hex())
+        .fetch_all(&database.pool)
+        .await?;
+
+        Ok(rows.into_iter().map(|r| r.into()).collect())
+    }
+
+    /// Finds all pending AccountGroups for a given account.
+    /// Pending means: user_confirmation is NULL.
+    pub(crate) async fn find_pending_for_account(
+        account_pubkey: &PublicKey,
+        database: &Database,
+    ) -> Result<Vec<Self>, sqlx::Error> {
+        let rows = sqlx::query_as::<_, AccountGroupRow>(
+            "SELECT id, account_pubkey, mls_group_id, user_confirmation, created_at, updated_at
+             FROM accounts_groups
+             WHERE account_pubkey = ? AND user_confirmation IS NULL",
+        )
+        .bind(account_pubkey.to_hex())
+        .fetch_all(&database.pool)
+        .await?;
+
+        Ok(rows.into_iter().map(|r| r.into()).collect())
+    }
+
+    /// Updates the user_confirmation status for this AccountGroup.
+    pub(crate) async fn update_user_confirmation(
+        &self,
+        user_confirmation: bool,
+        database: &Database,
+    ) -> Result<Self, sqlx::Error> {
+        let id = self.id.ok_or_else(|| {
+            sqlx::Error::Protocol("Cannot update unsaved AccountGroup".to_string())
```

Comment:
`sqlx::Error::Protocol` is kind of a weird error but I'm not sure which one would be better here, what do you think?